### PR TITLE
refactor intrinsic op

### PR DESCRIPTION
**Pull request by thewilsonator**
cc @ibuclaw

Since you were changing the intrinsic detection I thought I’d have one last crack at getting it to work like this.

Rebases #8942

### CODE REVIEW

**Comment by ibuclaw** [comment]
@thewilsonator - thanks, I always get mild irritation whenever I amend this list. ;-)

**Comment by thewilsonator** [comment]
Np problems. Next I suppose is the same thing in `builtin.d`. This is good to go with approval as it is green. Buildkite (or rather digitalmars.com) is having network issues, and Azure somehow got cancelled.

**Review by ibuclaw — APPROVED** [review]
Seems reasonable to me.

**Comment by rainers** [comment]
I agree this is easier to maintain than the list of mangled names and decoupled op arrays. It is dropping the check for function attributes and arguments, though. Is that unnecessary or can it cause trouble later on?

**Comment by thewilsonator** [comment]
> It is dropping the check for function attributes and arguments, though.

Attributes, yes. Arguments, sometimes, when there is no 64-bit version available (either outright, or because we're targeting a 32bit CPU).

> Is that unnecessary or can it cause trouble later on?

Modulo availability (described above), I think it is unnecessary. Also it should be easy to add back if for some reason it is determined to be important.

File: src/dmd/toir.d [hunk 1]
```diff
@@ -392,385 +392,115 @@ enum NotIntrinsic = -1;
...
-                op == core_ioptab[i - 1])
-            {
-                // Don't recognize 64 bit bsf() / bsr() in 32 bit mode
-                op = NotIntrinsic;
-            }
-            return op;
+        if (op == OPbsf || op == OPbsr ) {
+            //_D4core5bitop3bsfFNaNbNiNfxZi
+            // index of x == 26, m == ulong
+            if (mangleExact(fd)[26] == 'm')
```
Comment (rainers) [inline src/dmd/toir.d]: If the attributes of `core.bitop.bsf` change, this might read garbage.

File: src/dmd/toir.d [hunk 2]
```diff
@@ -392,385 +392,115 @@ enum NotIntrinsic = -1;
...
-        if (global.params.is64bit &&
-            fd.toParent().isTemplateInstance() &&
-            fd.ident == Id.va_start)
+Lva_start:
+    if (global.params.is64bit &&
+        fd.toParent().isTemplateInstance() &&
+        fd.ident == Id.va_start)
+    {
+        OutBuffer buf;
+        mangleToBuffer(fd.getModule(), &buf);
```
Comment (rainers) [inline src/dmd/toir.d]: Can't you use the caching `mangleExact` here, too? Redoing the mangling is a bit expensive.

File: src/dmd/toir.d (hunk 1 above)
Comment (thewilsonator) [reply src/dmd/toir.d]: Yeah I don't like that either, but its much faster than a `strcmp` and for some reason I couldn't get a type check on `fd` to do what I wanted: `const bool isFirstArgUlong = fd.parameters && (*fd.parameters)[0].type == Type.tuns64;`

File: src/dmd/toir.d (hunk 2 above)
Comment (thewilsonator) [reply src/dmd/toir.d]: That is just moving code that was already there. I'd love to get rid of it in a future PR though.

File: src/dmd/toir.d (hunk 1 above)
Comment (rainers) [reply src/dmd/toir.d]: This PR already avoids a lot of string comparisons, so I think adding one is fine here. You could implement a small local `endsWith` to skip the attribute checks.

File: src/dmd/toir.d [hunk 3]
```diff
@@ -392,385 +392,115 @@ enum NotIntrinsic = -1;
  */
 int intrinsic_op(FuncDeclaration fd)
 {
+    int op = NotIntrinsic;
     fd = fd.toAliasFunc();
-    const char *name = mangleExact(fd);
+    if (fd.isDeprecated())
```
Comment (rainers) [inline src/dmd/toir.d]: Is this additional check necessary?

**Comment by thewilsonator** [comment]
@rainers I'll fix those in another PR. No need to keep @ibuclaw waiting.

**Comment by MoonlightSentinel** [comment]
This broke `Azure pipelines (Windows_VisualD_Debug x64-debug-dmd)`

**Comment by thewilsonator** [comment]
https://github.com/dlang/dmd/pull/10660

**Comment by RazvanN7** [comment]
Maybe we should revert this until we come up with a solution? Having our PR pipeline stalled for so long is counter-productive.

**Comment by rainers** [comment]
> Maybe we should revert this until we come up with a solution? Having our PR pipeline stalled for so long is counter-productive.

The issues of this PR should be fixed by https://github.com/dlang/dmd/pull/10665. AFAICT the PR queue is plagued by networking issues.

**Comment by WalterBright** [comment]
This caused a regression: https://issues.dlang.org/show_bug.cgi?id=20446

**Pull request merged**